File: runtime/CSlib/input_port.cs
using System;

namespace bigloo
{
  public abstract class input_port: obj
  {
    public readonly String name;
    public int filepos = 0;
    public int pseudoeof = -1;
    public long length = -1;
    public int bufsiz;
    public bool eof = false;
    public bool other_eof = false;
    public int matchstart = 0;
    public int matchstop = 0;
    public int forward = 0;
    public byte lastchar = (byte)'\n';
    public int bufpos = 1;
    public byte[] buffer;
    public Object chook = bigloo.foreign.BUNSPEC;

    public input_port( String name, int bufsiz )
    {
      this.name= name;
      this.bufsiz= bufsiz;
      buffer= new byte[bufsiz];
    }

    public input_port( String name, byte[] buf )
    {
      this.name= name;
      this.bufsiz= buf.Length;
      buffer= buf;
    }

    public virtual void close()
    {
       if( chook is procedure )
       {
	  ((procedure)chook).funcall1(this);
       }
    }

    public abstract bool rgc_charready();

    public virtual bool reset_eof()
    {
      return false;
    }

    public abstract bool rgc_fill_buffer();

    public void rgc_enlarge_buffer_size( int  new_size )
    {
      int bufsize = this.bufsiz;

      if (new_size < bufsize)
        return;

      if (bufsize == 2)
        foreign.fail( "input-port",
                      "Can't enlarge buffer for non bufferized port (see the user manual for details)",
                      this );
      else {
        byte[] obuffer = buffer;
        byte[] nbuffer = new byte[new_size];

        for ( int i= 0 ; i < bufsize ; ++i )
          nbuffer[i]= obuffer[i];
        this.bufsiz= new_size;
        buffer= nbuffer;
      }
    }

    public void rgc_double_buffer()
    {
      rgc_enlarge_buffer_size( 2 * bufsiz );
    }

    public virtual object bgl_input_port_seek( int  pos )
    {
      return bbool.faux;
    }

    public virtual Object bgl_input_port_reopen()
    {
      return bbool.faux;
    }

    public override void write( output_port  p )
    {
      p.write( "[PORT " + name + " @" + filepos + "-" + matchstart + "." + forward + "." + matchstop + "-" + bufsiz + "." + bufpos + "]" );
    }
  }
}

File: runtime/Jlib/input_port.java
package bigloo;

import java.io.*;

public abstract class input_port extends obj
{
  public String name;
  public int filepos = 0;
  public int pseudoeof = -1;
  public long length = -1;
  public boolean eof = false;
  public boolean other_eof = false;
  public int matchstart = 0;
  public int matchstop = 0;
  public int forward = 0;
  public byte lastchar = (byte)'\n';
  public int bufpos = 0;
  public byte[] buffer;
  public Object chook = bigloo.foreign.BUNSPEC;
  public Object userseek = bigloo.foreign.BUNSPEC;

  public input_port( final String name, final byte[] buf )
  {
    this.name= name;
    buffer= buf;
  }

  public void close() 
  {
    if( chook instanceof procedure )
     {
	((procedure)chook).funcall1(this);
     }
  }

  public abstract boolean rgc_charready();

  public boolean reset_eof()
  {
    return false;
  }

  public abstract boolean rgc_fill_buffer()
    throws IOException;

   final void rgc_enlarge_buffer_size( int nsize ) {
    final int bufsize= this.buffer.length;

    if( nsize < bufsize ) return;
    
    if (bufsize == 2)
      foreign.fail( "input-port",
                    "Can't enlarge buffer for non bufferized port (see the user manual for details)",
                    this );
    else
    {
      final byte[] obuffer= buffer;
      final byte[] nbuffer= new byte[nsize];

      for ( int i= 0 ; i < bufsize ; ++i )
        nbuffer[i]= obuffer[i];
      buffer= nbuffer;
    }
  }

   public final void rgc_double_buffer() {
      rgc_enlarge_buffer_size( this.buffer.length * 2 );
   }
   
  Object bgl_input_port_seek( final int pos )
    throws IOException
  {
    return bigloo.foreign.BFALSE;
  }

  Object bgl_input_port_reopen() throws IOException {
    return bigloo.foreign.BFALSE;
  }

  public Object bgl_input_port_clone( input_port src ) {
     filepos = src.filepos;
     pseudoeof = src.pseudoeof;
     length = src.length;
     eof = src.eof;
     other_eof = src.other_eof;
     matchstart = src.matchstart;
     matchstop = src.matchstop;
     forward = src.forward;
     lastchar = src.lastchar;
     bufpos = src.bufpos;
     buffer = src.buffer;
     chook = src.chook;

     return this;
  }

  public void write( final output_port p )
  {
     p.write( "[PORT " + name + " @" + filepos + "-" + matchstart + "." + forward + "." + matchstop + "-" + buffer.length + "." + bufpos + "]" );
  }

   public boolean timeout_set( int to ) {
      return false;
   }
}
